Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -5,182 +5,289 @@
+  forM_ base $ \(prime, roots) -> case someNatVal (intToNatural (unPrime prime)) of
+    SomeNat (Proxy :: Proxy prime) -> do
+      let
+        startingIndices = case MW.invertMod (fromInteger a :: MW.Mod prime) of
+          Just inverseOfA -> map (\root -> (wordToInt . MW.unMod) (fromIntegral m + fromInteger (- b + root) * inverseOfA :: MW.Mod prime)) roots
+          Nothing         -> case MW.invertMod (fromInteger (2 * b) :: MW.Mod prime) of
+            -- Temporary fix. If power > 2, something better has to be found.
+            Just inverseOf2B -> [(wordToInt . MW.unMod) (fromIntegral m - fromInteger c * inverseOf2B :: MW.Mod prime)]
+            -- For this to be true 2 cannot be a factor of a.
+            -- Better to ouput @unPrime prime@ as a factor.
+            Nothing          -> error ("Found an illegal factor: " ++ show prime)
+      forM_ startingIndices $ \startingIndex -> do
+        let change y = y - (intLog2 . unPrime) prime
+        forM_ [startingIndex, startingIndex + unPrime prime..(2 * m)] $ \entry ->
+          MV.modify sievingIntervalM change entry
 
 -- This algorithm filters @sievingIntervalF@ for smooth numbers. It returns
 -- the smooth numbers together with their @index@ and @startingPoint@.
 -- This is needed in order to later compute @firstSquare@.
-findSmoothNumbers :: Integer -> V.Vector (Integer, SignedPrimeIntSet) -> V.Vector (Integer, SignedPrimeIntSet)
-findSmoothNumbers startingPoint = V.imapMaybe selectSmooth
+findLogSmoothNumbers :: [Prime Int] -> Int -> Int -> [(Prime Integer, Word)] -> Integer -> V.Vector (Integer, Int) -> V.Vector (Integer, I.IntMap Int)
+findLogSmoothNumbers factorBase m h decompositionOfA b = V.imapMaybe selectSmooth
   where
-    selectSmooth index (residue, factorisation)
-      | residue == 1 = Just (intToInteger index + startingPoint, factorisation)
-      | otherwise    = Nothing
+    a = factorBack decompositionOfA
+    intMapA = I.fromAscList $ map (bimap (integerToInt . unPrime) wordToInt) decompositionOfA
+    highestPrime = intToInteger . unPrime . last $ factorBase
+    selectSmooth index (value, logResidue) = case logResidue < h of
+      True
+        -- Cannot use factorisation instead of listFactorisation since its rightmost value need not be an int.
+        | null listFactorisation                         -> Just (a * intToInteger (index - m) + b, factorisation)
+        | (fst . last) listFactorisation <= highestPrime -> Just (a * intToInteger (index - m) + b, factorisation)
+        | otherwise                                      -> Nothing
+      False -> Nothing
+      where
+        factorisation = I.unionWith (+) intMapA $ if value < 0 then I.insert (-1) 1 preFac else preFac
+        preFac = I.fromAscList $ map (bimap integerToInt wordToInt) listFactorisation
+        -- Maybe there is a better way to use @trialDivision@
+        listFactorisation = trialDivisionWith (map (intToInteger . unPrime )factorBase) value
```

Comment:
Is `trialDivisionWith` taking a significant time? There are ways to speed up it a bit.

Reply:
It can take some time if the smoothness threshold is high. I think it's worth trying to speed it up